src/Database/PostgreSQL/Consumers/Config.hs

```diff
@@ -102,9 +92,51 @@ data ConsumerConfig m idx job = forall row. FromRow row => ConsumerConfig {
 -- | Function that processes a job. It's recommended to process each
 -- job in a separate DB transaction, otherwise you'll have to remember
 -- to commit your changes to the database manually.
-, ccProcessJob            :: !(job -> m Result)
+, ccProcessJob            :: !(Job idx info -> m Result)
 -- | Action taken if a job processing function throws an exception.
 -- Note that if this action throws an exception, the consumer goes
 -- down, so it's best to ensure that it doesn't throw.
-, ccOnException           :: !(SomeException -> job -> m Action)
+, ccOnException           :: !(SomeException -> Job idx info -> m Action)
 }
+
+data Job idx info = Job {
+  idx        :: idx
+, runAt      :: Maybe UTCTime
+, finishedAt :: Maybe UTCTime
+, attempts   :: Int
+, info       :: info
+}
+
+-- | Result of processing a job.
+data Result = Ok Action | Failed Action
+  deriving (Eq, Ord, Show)
+
+-- | Action to take after a job was processed.
+data Action
+  = MarkProcessed
+  | RerunAfter Interval
+  | RerunAt UTCTime
+  | Remove
+    deriving (Eq, Ord, Show)
+
+exponentialBackoff :: (Monad m)
+  => Int -> Interval -> SomeException -> Job idx info -> m Action
+exponentialBackoff maxAttempts startInterval _ Job{..} = return $
+  case (attempts > maxAttempts, attempts > 2) of
+    (True, _)      -> Remove
+    (False, False) -> RerunAfter startInterval
+    (False, True)  -> RerunAfter $
+      diffTimeToInterval $ intervalToDiffTime startInterval ^ attempts
```

Comment:
and another mistake here, to get exponential backoff behaviour I believe it should actually be:
```hs
(intervalToDiffTime startInterval) * 2 ^ attempts
```